### PR TITLE
add legacy load utility

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Splits changes off #8558 

Adds a context manager to handle legacy checkpoints that have pickled attributes in the content that was removed from lightning in new versions. The currently known ones are:

- `pytorch_lightning.utilities.argparse_utils` (renamed, slated for removal)
- `pytorch_lightning.utilities.argparse._gpus_arg_default` (dead code)

We can remove these dead code pieces and instead dynamically patch modules to re-route the imports. 
The following works:

```python
    with pl_legacy_patch():
        # this would normally result in ImportError
        from pytorch_lightning.utilities.argparse import _gpus_arg_default
```

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

### CODE REVIEW

**Comment by awaelchli** [comment]
@carmocca the main motivation of this standalone PR here is to allow us to remove the dead code files which we otherwise would need to keep forever. But as you said it also ties in with any automatic upgrading we might do in the future where we may need to patch out attributes before we can apply a migration. I think a context manger is the most convenient way forward, unless we say at a certain point we drop support for loading old checkpoints, then we would not need it anymore.

**Comment by carmocca** [comment]
>  I think a context manger is the most convenient way forward

I thought we would do the incremental upgrades as functional transforms over the loaded data. 

```python
data = load(ckpt_file)
data = upgrade_to_1_0_0(data)
data = upgrade_to_...(data)
return data
```

or something similar.

Although as I understand it, this wouldn't work for the particular case of `_gpus_arg_default` because the loading itself would fail, correct?

So we would do

```python
with pl_legacy_patch():
    data = load(ckpt_file)
data = upgrade_to_1_0_0(data)
data = upgrade_to_...(data)
return data
```

**Comment by awaelchli** [comment]
Yes exactly, nothing has changed in this approach.

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/utilities/migration.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+           but still needs to be available for import for legacy checkpoints.
+
+    Example:
+
+        with pl_legacy_patch():
+            torch.load("path/to/legacy/checkpoint.ckpt")
+    """
+
+    def __enter__(self):
+        # `pl.utilities.argparse_utils` was renamed to `pl.utilities.argparse`
```
Comment (tchaton) [inline pytorch_lightning/utilities/migration.py:37]: Neat !

File: pytorch_lightning/core/saving.py [hunk 2]
```diff
@@ -125,10 +126,11 @@ def load_from_checkpoint(
             pretrained_model.freeze()
             y_hat = pretrained_model(x)
         """
-        if map_location is not None:
-            checkpoint = pl_load(checkpoint_path, map_location=map_location)
-        else:
-            checkpoint = pl_load(checkpoint_path, map_location=lambda storage, loc: storage)
+        with pl_legacy_patch():
+            if map_location is not None:
+                checkpoint = pl_load(checkpoint_path, map_location=map_location)
```
Comment (tchaton) [inline pytorch_lightning/core/saving.py:131]: @awaelchli quick question: would see any value to bring this context manager directly inside pl_load ?

**Review by tchaton — APPROVED** [review]

File: pytorch_lightning/core/saving.py (hunk 2 above)
Comment (awaelchli) [reply pytorch_lightning/core/saving.py:131]: yes, there may be a minor value, however I think we might want to keep the ones in the checkpoint connector anyway since there we load using the training type plugin / checkpointIO plugin and if a custom checkpoint io plugin does not call pl_load it would need to remember calling `pl_legacy_patch`

**Comment by awaelchli** [comment]
> shall include such case to testing of legacy checkpoints?

@Borda sorry I missed your question. 
Are you suggesting to add a 1.2.8 legacy checkpoint to https://pl-public-data.s3.amazonaws.com/legacy/checkpoints.zip ?

Of course, I made sure that the legacy patch works with the old checkpoints. Below is the code I used to verify it. 
I can provide the checkpoint file and we can make a test for it but seems a bit overkill imo.


```python

import torch
from argparse import ArgumentParser
from torch.utils.data import Dataset

from pytorch_lightning import LightningModule, Trainer


class RandomDataset(Dataset):
    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self):
        return self.len


class BoringModel(LightningModule):
    def __init__(self, **kwargs):
        super().__init__()
        self.save_hyperparameters(kwargs)
        self.layer = torch.nn.Linear(32, 2)

    def forward(self, x):
        return self.layer(x)

    def loss(self, batch, prediction):
        # An arbitrary loss to have a loss that updates the model weights during `Trainer.fit` calls
        return torch.nn.functional.mse_loss(prediction, torch.ones_like(prediction))

    def step(self, x):
        x = self.layer(x)
        out = torch.nn.functional.mse_loss(x, torch.ones_like(x))
        return out

    def training_step(self, batch, batch_idx):
        output = self.layer(batch)
        loss = self.loss(batch, output)
        return {"loss": loss}

    def training_step_end(self, training_step_outputs):
        return training_step_outputs

    def training_epoch_end(self, outputs) -> None:
        torch.stack([x["loss"] for x in outputs]).mean()

    def configure_optimizers(self):
        optimizer = torch.optim.SGD(self.layer.parameters(), lr=0.1)
        lr_scheduler = torch.optim.lr_scheduler.StepLR(optimizer, step_size=1)
        return [optimizer], [lr_scheduler]


def run_1_2_7():
    # RUN WITH LIGHTNING 1.2.7

    # fake data
    train_data = torch.utils.data.DataLoader(RandomDataset(32, 64))

    # model
    parser = ArgumentParser()
    parser = Trainer.add_argparse_args(parser)
    args = parser.parse_args()
    trainer = Trainer.from_argparse_args(args, default_root_dir="legacy_logs1.2.7", max_steps=1)
    model = BoringModel(**vars(args))
    trainer.fit(model, train_data)
    print(trainer.checkpoint_callback.best_model_path)


def run_master():
    # RUN WITH LIGHTNING 1.2.8
    from pytorch_lightning.utilities.migration import pl_legacy_patch

    with pl_legacy_patch():
        # can unpickle!
        x = torch.load("legacy_logs1.2.7/lightning_logs/version_2/checkpoints/epoch=0-step=0.ckpt")
        assert callable(x["hyper_parameters"]["gpus"])


if __name__ == "__main__":
    run_master()

```

**Review by Borda — APPROVED** [review]
great solution and pattern for future similar changes!
@awaelchli mind resolve conflicts? :rabbit:
